### ISSUE TITLE
Stream.CloseAsync method

### ISSUE THREAD

**Issue by tmds** [dotnet/corefx#42704]
I've had the need for a method which isn't available on `Stream`:

```cs
ValueTask CloseAsync(CancellationToken);
```

This method would allow to implement close operation that can be asynchronous and cancelled+aborted.

The two existing methods for this on `Stream` are:
```cs
void Close (); // not async
ValueTask DisposeAsync (); // no cancellation token
```
`Close` is not async, and `DisposeAsync` doesn't accept a cancellation token.

Cancelling the `CloseAsync` token has abort semantics. That is, unflushed data is discarded, and for example `NetworkStream` would perform a RST close (lingertime 0) on the Socket.

I wonder if it would be a meaningful addition to the `Stream` class.

@dotnet/ncl @stephentoub @geoffkizer @davidsh @davidfowl @AArnott @benaadams

**Comment by stephentoub** [comment]
I don't think such a method should be added.

If Close is provided on an IDisposable, it is expected that Close and Dispose are functionally the same. In the case of Stream, Dispose is even non-virtual and just calls Close. So I think it would be very unintuitive if there were a CloseAsync that was functionally different from DisposeAsync.

Further, we explicitly chose not to include a CancellationToken on DisposeAsync because DisposeAsync is primarily about cleaning up resources, and such resources should be cleaned up even in the face of cancellation, e.g. it'd be problematic if code like:
```C#
try
{
    Foo(stream, token);
}
finally
{
    await stream.DisposeAsync(token);
}
```
didn't clean up after the stream if the token had cancellation requested.  And while you could say "then don't pass the token to DisposeAsync", that's also unintuitive, and then gets back to why would DisposeAsync take a token in the first place.

**Comment by tmds** [comment]
>If Close is provided on an IDisposable, it is expected that Close and Dispose are functionally the same. In the case of Stream, Dispose is even non-virtual and just calls Close. So I think it would be very unintuitive if there were a CloseAsync that was functionally different from DisposeAsync.

ok, maybe Close is not a good name then.

> DisposeAsync is primarily about cleaning up resources

Yes. The method I'm looking for is to implement the protocol part in a way that is cancellable.
resource cleanup should be in `DisposeAsync` and not cancellable.

```cs
async using stream = ...;
await stream.CloseAsync(token); // protocol close
```

**Comment by tmds** [comment]
Another API option could be to have a boolean that disables flushing/protocol close that is implemented in `DisposeAsync`.
```cs
bool AbortiveClose { set; }
```

**Comment by scalablecory** [comment]
I think there's something here. Graceful close and abortive close (dispose) are two separate concepts. There are many protocols with a graceful close feature.

**Comment by AArnott** [comment]
> There are many protocols with a graceful close feature.

True. .NET's own `WebSocket` being one of them.

**Comment by stephentoub** [comment]
Yes. But it's not exposed through Stream. If "close" is the terminology used in the protocol (as it is with WebSocket) it's reasonable to expose an API on the concrete implementation of that protocol with that name (as it is with WebSocket).

**Comment by stephentoub** [comment]
Just to be clear, my main concern is confusion between close and dispose. If we think we can come up with a way to avoid confusion, we could consider it.

cc: @bartonjs, @KrzysztofCwalina, @terrajobst

**Comment by scalablecory** [comment]
It would be mildly infuriating to have yet another similarly-sounding verb, but a `ShutdownAsync(CancellationToken)` would be nice.
